core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java

```diff
@@ -134,50 +138,81 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
-                return new BaseAddedRowsScanTask(
-                    changeOrdinal,
-                    commitSnapshotId,
-                    dataFile,
-                    NO_DELETES,
-                    context.schemaAsString(),
-                    context.specAsString(),
-                    context.residuals());
-
-              case DELETED:
-                return new BaseDeletedDataFileScanTask(
-                    changeOrdinal,
-                    commitSnapshotId,
-                    dataFile,
-                    NO_DELETES,
-                    context.schemaAsString(),
-                    context.specAsString(),
-                    context.residuals());
-
-              default:
-                throw new IllegalArgumentException("Unexpected entry status: " + entry.status());
-            }
-          });
+      return CloseableIterable.filter(
+          CloseableIterable.transform(
+              entries,
+              entry -> {
+                long snapshotId = entry.snapshotId();
+                Snapshot snapshot = addedToChangedSnapshots.get(snapshotId);
+                long commitSnapshotId = snapshot.snapshotId();
+                int changeOrdinal = snapshotOrdinals.get(snapshot.snapshotId());
+                DataFile dataFile = entry.file().copy(context.shouldKeepStats());
+                DeleteFile[] deleteFiles = context.deletes().forDataFile(dataFile);
+                List<DeleteFile> addedDeletes = Lists.newArrayList();
+                List<DeleteFile> existingDeletes = Lists.newArrayList();
+                for (DeleteFile file : deleteFiles) {
+                  if (file.dataSequenceNumber() == snapshot.sequenceNumber()) {
+                    addedDeletes.add(file);
+                  } else {
+                    existingDeletes.add(file);
+                  }
+                }
+
+                switch (entry.status()) {
+                  case ADDED:
+                    if (snapshotId == commitSnapshotId) {
+                      return new BaseAddedRowsScanTask(
+                          changeOrdinal,
+                          commitSnapshotId,
+                          dataFile,
+                          addedDeletes.toArray(new DeleteFile[0]),
+                          context.schemaAsString(),
+                          context.specAsString(),
+                          context.residuals());
+                    } else if (deleteFiles.length > 0) {
+                      return new BaseDeletedRowsScanTask(
+                          changeOrdinal,
+                          commitSnapshotId,
+                          dataFile,
+                          addedDeletes.toArray(new DeleteFile[0]),
+                          existingDeletes.toArray(new DeleteFile[0]),
+                          context.schemaAsString(),
+                          context.specAsString(),
+                          context.residuals());
+                    } else {
+                      return null;
```

Comment:
That is one way! ;-)
In my implementation, I use a placeholder `DummyChangelogScanTask` that is filtered out.